packages/lib/src/vis/tiles/TiledHeatmap.tsx

```diff
@@ -0,0 +1,68 @@
+import { clamp, range } from 'lodash';
+
+import { getInterpolator } from '../heatmap/utils';
+import { useFrameRendering } from '../hooks';
+import { useAxisSystemContext } from '../shared/AxisSystemContext';
+import TiledHeatmapLayer from './TiledHeatmapLayer';
+import type { TilesApi } from './api';
+import { useDataPerPixel } from './hooks';
+import type { ColorMapAttributes } from './models';
+
+interface Props extends ColorMapAttributes {
+  api: TilesApi;
+  displayLowerResolutions?: boolean;
+  qualityFactor?: number;
+}
+
+function TiledHeatmap(props: Props) {
+  const {
+    api,
+    displayLowerResolutions = true,
+    qualityFactor = 1, // 0: Lower quality, less fetch; 1: Best quality
+    ...colorMapAttrs
+  } = props;
+
+  const { visSize } = useAxisSystemContext();
+  const { xDataPerPixel, yDataPerPixel } = useDataPerPixel();
+
+  const nLevels = api.levelsSize.length;
+
+  const quality = 1 - clamp(qualityFactor, 0, 1);
+  const levelOfDetail = Math.min(
+    Math.floor(Math.log2(Math.max(1, xDataPerPixel, yDataPerPixel)) + quality),
+    Math.max(nLevels - 1, 0)
+  );
```

Comment:
This computation goes a little bit over my head. Any way to split it with additional, self-descriptive variables and/or explain what's going on with a comment or two?

Can we assume that `nLevels` is greater than 0 to avoid the second `Math.max`? You could put a little assertion in the constructor of `TileApi` to check for a minimum length (`if() { throw new Error('Expected `levelSizes` to have at least one size') }`). Alternatively, the constructor could accept the image size as a separate parameter, and could prepend it to the `levelSizes` array -- this would guarantee a min length of 1. Or you could keep the image size separate from the other levels and just do `[api.imageSize, ...levelSizes]` where you need to loop over all the sizes.

Reply:
Hopefully made more explicit and checking for empty array.